service/src/event/ordering_task.rs

```diff
@@ -175,302 +97,470 @@ impl OrderingTask {
+        new.add_event(event);
+        new
     }
 
-    /// returns Some(Stream Init CID) if this is a new event, else None.
-    pub fn add_event(&mut self, event: DeliverableEvent) -> Option<InitCid> {
-        let res = if self.prev_map.insert(event.meta.prev, event.cid).is_none() {
-            Some(event.meta.init_cid)
-        } else {
-            None
+    /// returns true if this is a new event.
+    fn add_event(&mut self, event: StreamEvent) -> bool {
+        let cid = match &event {
+            StreamEvent::InitEvent(cid) => {
+                self.update_should_process_for_new_delivered(cid);
+                *cid
+            }
+            StreamEvent::KnownDeliverable(meta) => {
+                self.prev_map.insert(meta.prev, meta.cid);
+                self.update_should_process_for_new_delivered(&meta.cid);
+                meta.cid
+            }
+            StreamEvent::Undelivered(meta) => {
+                self.prev_map.insert(meta.prev, meta.cid);
+                if !self.should_process {
+                    // we depend on something in memory
+                    if let Some(in_memory_prev) = self.cid_map.get(&meta.prev) {
+                        match in_memory_prev {
+                            StreamEvent::InitEvent(_) | StreamEvent::KnownDeliverable(_) => {
+                                self.should_process = true;
+                            }
+                            StreamEvent::Undelivered(_) => {
+                                // nothing to do until it's prev arrives
+                            }
+                        }
+                    }
+                }
+                meta.cid
+            }
         };
-        self.cid_map.insert(event.cid, event);
-        res
+
+        self.cid_map.insert(cid, event).is_none()
     }
 
-    pub fn is_empty(&self) -> bool {
-        // these should always match
-        self.prev_map.is_empty() && self.cid_map.is_empty()
+    /// we'll be processed if something in memory depends on this event
+    fn update_should_process_for_new_delivered(&mut self, new_cid: &EventCid) {
+        // don't reset the flag if we're already supposed to be processed
+        if !self.should_process {
+            self.should_process = self.prev_map.contains_key(new_cid);
+        }
     }
 
-    fn remove_by_event_cid(&mut self, cid: &Cid) -> Option<DeliverableEvent> {
-        if let Some(cid) = self.cid_map.remove(cid) {
-            self.prev_map.remove(&cid.meta.prev);
-            Some(cid)
-        } else {
-            None
+    /// Called when we've persisted the deliverable events to the database and can clean up our state.
+    /// Returns `true` if we're finished processing and can be dropped from memory.
+    /// Returns `false` if we have more work to do and should be retained for future processing
+    fn processing_completed(&mut self) -> bool {
+        self.should_process = false;
+        self.new_deliverable.clear();
```

Comment:
should we remove entries from `cid_map` and `prev_map` based on what was in new_deliverable?  No need to keep maintaining in-memory state for events that were fully delivered.

It's theoretically possible a stream could keep having new events marked deliverable without ever having *all* of its events deliverable all at once so the state never gets dropped and the amount of data being stored related to old already delivered events just grows and grows, but I think we can do something here to prune what we're keeping in memory and avoid that.

Reply:
I don't want to make this change now but added this comment.

```
        // It's worth considering whether we should prune the `cid_map` and `prev_map` with everything from the `new_deliverable` set to avoid
        // memory leaks due to never finding events. This requires setting `self.should_process` to true for all undelivered events (or making a db query),
        // as we will no longer have everything we've discovered in memory and need to go to the db to figure out what to do. As streams are generally
        // short, I'm opting to not make this change now. We could consider something like an expiration time as well, but generally we don't want
        // to be dropping things from memory that we might need to process in the future unless we have to.
```

Reply:
I don't understand that comment, that seems way more complicated than I was imagining.  Why do we need to go back to the database?  Can't we just loop once over `new_deliverable` and if any of those events are keys in `cid_map` or values in `prev_map`, just remove them from the map?

Reply:
Also the memory leak isn't due to "never finding events".  Even if you always find all events eventually you can wind up with a memory leak if by the time you've filled one hole a new hole has always shown up so that the full `StreamEvents` struct never gets dropped.

For example imagine a stream where you get delivered events in the following pattern:
1, 3, 5, 2, 7, 4, 9, 6, 11, etc.  You never have a moment where you have all the events in the stream without any holes, but you also do get every event eventually and there is always a point in the stream after which you have received everything and there are no holes earlier than that point, which should allow you to clean up any state associated with those earlier events that are before the point where holes remain.

Reply:
Yes, we can do that and it works fine as long as we make sure to be aggressive about looking up history for the messages we're discovering. The problem is that messages can sit in the channel so there's a period after they are checked for being deliverable initially and when they get reviewed, which can cause things to change. Currently, by keeping the history for partially processed streams, we're able to see what's happened while we've been in the queue. 

But, this isn't totally true. Even though I [just changed this](https://github.com/ceramicnetwork/rust-ceramic/pull/387/commits/e4f5a6e8d008afce02ebb609244a07dd6d700ae5) to try to avoid extra work, it might be better to set for every undelivered event, since I'm imagining a forking scenario when the queue is deep that results in our history being updated and dropped, then some new fragment we don't depend on is in memory. In that case, relying on the recent changes in memory won't get us reviewed, so it seems better to do extra work and make sure we don't miss anything for now.

Anyway, with the change to always make sure to process every undelivered event, pruning works fine by looping through and removing things, but it will result in more database lookups since we'll be dropping things we're about to need, especially when the queue is deeper.

Reply:
We talked, was so much easier 😅.. made the change in [this commit](https://github.com/ceramicnetwork/rust-ceramic/pull/387/commits/1c0ca5fa35121af4f8dd773dbf5f91763ca3a4ec).